src/System.Text.Encodings.Web/src/System/Text/Internal/AllowedCharactersBitmap.cs

```diff
@@ -85,18 +82,54 @@ public bool IsUnicodeScalarAllowed(int unicodeScalar)
         {
             int index = unicodeScalar >> 5;
             int offset = unicodeScalar & 0x1F;
-            return ((_allowedCharacters[index] >> offset) & 0x1U) != 0;
+            return (_allowedCharacters[index] & (0x1U << offset)) != 0;
         }
 
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public unsafe int FindFirstCharacterToEncode(char* text, int textLength)
         {
-            for (int i = 0; i < textLength; i++)
+            int i = 0;
+
+            while (i < textLength - 8)
+            {
+                if (!IsCharacterAllowed(text[i])
+                    || !IsCharacterAllowed(text[++i])
+                    || !IsCharacterAllowed(text[++i])
+                    || !IsCharacterAllowed(text[++i])
+                    || !IsCharacterAllowed(text[++i])
+                    || !IsCharacterAllowed(text[++i])
+                    || !IsCharacterAllowed(text[++i])
+                    || !IsCharacterAllowed(text[++i]))
+                {
+                    goto Return;
+                }
+                i++;
+            }
```

Comment:
This loop has quite a long read-after-write-dependency on `i`. Maybe it's better to write this as
```c#
for (; i < textLength - 8; i += 8)
{
    if (!IsCharacterAllowed(text[i])
        || !IsCharacterAllowed(text[i + 1])
        || !IsCharacterAllowed(text[i + 2])
        || !IsCharacterAllowed(text[i + 3])
        || !IsCharacterAllowed(text[i + 4])
        || !IsCharacterAllowed(text[i + 5])
        || !IsCharacterAllowed(text[i + 6])
        || !IsCharacterAllowed(text[i + 7]))
    {
        goto Return;
    }
}
```

Same for the 4-unrolled loop.

Reply:
Would this matter?

Reply:
Updated my comment. The suggested approach won't work.

Reply:
Ahhh, sorry I was too eager at looking at indexes only.